sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AvroUtils.java

```diff
@@ -73,8 +69,19 @@ public static Object convertAvroFormat(Field beamField, Object value) {
       default:
         throw new RuntimeException("Does not support converting unknown type value");
     }
+  }
 
-    return ret;
+  private static ReadableInstant safeToMillis(Object value) {
+    long subMilliPrecision = ((long) value) % 1000;
```

Comment:
How do we know that that the value returned is not millis but micro?

Reply:
It is because this `AvroUtils` is part of the BigQuery connector, not general. And we know BigQuery is always micros. Is that your question?

Reply:
Yes. I was asking because input is an object.

Reply:
I was also wondering where long is generated. If BQ timestamp converts to long which contains micros, will that lose precision?

I did some math:

BigQuery doc says TIMESTAMP type range is `0001-01-01 00:00:00 to 9999-12-31 23:59:59.999999 UTC`

so it should ranges from `-6.127488×10¹⁶ to 2.4976512×10¹⁷` in micros, calculated by `((0001-01-01 / 9999-12-31) -/+ 1970) * 12 * 30 * 24 *3600 * 1000 * 1000`, which falls into INT64's range. 

So no precision loss.